native/common/jp_gc.cpp

```diff
@@ -0,0 +1,182 @@
+#include <Python.h>
+#include "jpype.h"
+#include "pyjp.h"
+#include "jp_reference_queue.h"
+
+#ifdef WIN32
+#define USE_PROCESS_INFO
+#include <Windows.h>
+#include <psapi.h>
+#elif __APPLE__
+#define USE_TASK_INFO
+#include <unistd.h>
+#include <sys/resource.h>
+#include <mach/mach.h>
+#else
+// Linux doesn't have an available rss tally so use mallinfo
+#define USE_MALLINFO
+#include <malloc.h>
+#endif
+#define DELTA_LIMIT 10*1024*1024l
+#define HARD_LIMIT 200*1024*1024l
+
+namespace
+{
+}
+
+size_t getWorkingSize()
+{
+	size_t current = 0;
+#if defined(USE_PROCESS_INFO)
+	PROCESS_MEMORY_COUNTERS pmc;
+	GetProcessMemoryInfo(GetCurrentProcess(), &pmc, sizeof (pmc));
+	current = (size_t) pmc.WorkingSetSize;
+
+#elif defined(USE_TASK_INFO)
+	struct mach_task_basic_info info;
+	mach_msg_type_number_t count = MACH_TASK_BASIC_INFO_COUNT;
+	if (task_info(mach_task_self(), MACH_TASK_BASIC_INFO, (task_info_t) & info, &count) == KERN_SUCCESS)
+		current = (size_t) info.resident_size;
+	return (size_t) 0; /* query failed */
+
+#elif defined(USE_MALLINFO)
+	struct mallinfo mi;
+	mi = mallinfo();
+	current = (size_t) mi.uordblks;
```

Comment:
uordblks contains the number of free memory chunks, I thought you want to obtain use memory blocks instead?
I think this sort of memory information platform thingy has already been solved elsewhere (header only at best).

Reply:
uordblks  - The total number of bytes used by in-use allocations.

http://man7.org/linux/man-pages/man3/mallinfo.3.html

I looked and couldn't find much on this topic that was useful.  Lots of stuff pointing to /proc which required a ton of parsing but very little not much in the way of actual guidance.  If you have a header that addresses it then I can include it.

Reply:
As you said it seems like a hard task to do it in a portable fashion. I do not have a library at hand. At least we should ensure that the info is approximately the right thing and I'm not yet sure about it.

Reply:
You're right about uordblks

Reply:
I used a few random sources since I could not find anything definitive.  Do you happen to know if mallinfo functions on osx  as it would be best it we had a tested pattern.

chip.org/wiki/resident_set_size
https://stackoverflow.com/questions/38490320/how-to-query-amount-of-allocated-memory-on-linux-and-osx/43126663

My conclusion is that linux only gives the maximum rss in getrusage() which is okay if I don't mind it overflowing and then having to change my guess rate.  But that would be horrible for a program that has random behavior and so not really acceptable.  Once the high water hit the hard limit we have to call every time.  Thus a switch to mallinfo.  mallinfo apparently doesn't capture everything but it does get the majority of numpy and python.  

Windows page unfortunately gets both python and Java so the behavior is really odd.  Both languages are GC and they like holding on to pages so we get gross block changes rather than anything fine grain.   So the trigger is haphazard because we only see page level changes.  It keeps things under the we are leaking like crazy line, but a better predictor would help.  I am worried in real code that stuff staying in blocks will cause overall growth.  The alternative is to use HeapWalk, but that looked miserable and it doesn't give me much info.

OSX.... er don't have access so no clue what I am doing on that one.

I tried neutral methods at first by instrumenting the malloc/free call using Python and the hidden malloc size, but their auditing tools are not called until after the memory is freed so I can't get the free size and it is pretty dangerous for portability.  But it would have been the most direct as it would be locked to space managed by Python (and numpy).